drivers: flash : Cadence nand driver

nbalabak opened the pull request:
Implemented the Cadence NAND driver with the following features:

- Added new driver for Cadence Nand
- Supports 3 modes : CDMA , PIO, generic mode
- Support Operation : read, write, erase
- Supports x8 and x16
- Supports SDR datarate

nbalabak committed:
soc: arm64: intel_socfpga: changes in system_manager

SOCFPGA_SYSMGR_REG_BASE base address now read from Device tree

This commit changes the way the SOCFPGA_SYSMGR_REG_BASE base address is
determined. Previously, the address was hard-coded in the system_manager
source file. This commit changes the code to read the address from the
Device tree. This makes the code more flexible and allows the base
address to be different for different boards.

Signed-off-by: Navinkumar Balabakthan <navinkumar.balabakthan@intel.com>

nbalabak committed:
drivers: flash: Added cdns Nand Driver

Added Cadence NAND driver to support reading, erasing and writing data.

Signed-off-by: Navinkumar Balabakthan <navinkumar.balabakthan@intel.com>

nbalabak committed:
samples: drivers: Added Nand Test Application

Nand Test Application has added in sample/drivers folder.

Signed-off-by: Navinkumar Balabakthan <navinkumar.balabakthan@intel.com>

nbalabak committed:
CODEOWNERS: Add owner update

Code owner added for nandflash driver.

Signed-off-by: Navinkumar Balabakthan <navinkumar.balabakthan@intel.com>

nbalabak committed:
dts: arm64: intel: dtsi support for cadence Nand driver for Agilex5

dtsi support for nand controller added to bring up nand driver on Agilex5

Signed-off-by: Navinkumar Balabakthan <navinkumar.balabakthan@intel.com>

nbalabak commented:
Hello @de-nordic ,
               could you please run compliance check and scan code for license?

nbalabak commented:
Hello @de-nordic ,
could you please run compliance check and scan code for license?

nbalabak commented:
@kartben @SgrrZhf @povergoing could you please review nand driver?

povergoing reviewed (commented):
LGTM overall, but I am not a nand expert, only some trivial comments

de-nordic commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1617 @@
+	page_count = end_page_number - start_page_number;
+	page_count++;
+	if ((check_page_last == 1) && (check_page_first == 1)) {
+#if CONFIG_CDNS_NAND_CDMA_MODE
+		ret = cdns_nand_cdma_read(params, start_page_number, (char *)buffer, page_count);
+#elif CONFIG_CDNS_NAND_PIO_MODE
+		ret = cdns_nand_pio_read_write(params, start_page_number, (char *)buffer,
+					       page_count, CDNS_READ);
+#elif CONFIG_CDNS_NAND_GENERIC_MODE
+		ret = cdns_nand_gen_read_write(params, start_page_number, (char *)buffer,
+					       page_count, CDNS_READ);
+#endif
```
This blocks keep on repeating.
Wouldn't it look better if function would be defined, maybe inline, that would change depending on the config defined and called here?

nbalabak replied:
@de-nordic i created inline function and updated code

nbalabak commented on drivers/flash/Kconfig.cadence_nand:
```diff
@@ -0,0 +1,53 @@
+	prompt "Set the NAND Operating mode"
+	default CDNS_NAND_CDMA_MODE
+	help
+		Specify the Operating mode used by the driver.
```
@povergoing i updated changed

nbalabak commented on drivers/flash/Kconfig.cadence_nand:
```diff
@@ -0,0 +1,53 @@
+	int "Set the page count for a single transfer in the CDMA Mode"
+	default 10
+	help
+		Configure the page count for a single transfer in the CDMA Mode
```
@povergoing i updated changed

nbalabak commented on drivers/flash/Kconfig.cadence_nand:
```diff
@@ -0,0 +1,53 @@
+	int "Set the block count for a single transfer in the CDMA Mode"
+	default 10
+	help
+		Configure the block count for a single transfer in the CDMA Mode
```
@povergoing i updated changes

nbalabak replied:
@povergoing  Can you please review the changes I made based on your comments?

nbalabak commented on CODEOWNERS:
```diff
@@ -506,3 +509,339 @@
 /dts/bindings/gpio/*ads114s0x*            @benediktibk
 /dts/bindings/pwm/*max31790*              @benediktibk
 /dts/bindings/dac/*ad56*                  @benediktibk
+/dts/common/                              @galak
```
@de-nordic i updated changes

de-nordic commented on drivers/flash/flash_cadence_nand.c:
```diff
@@ -0,0 +1,292 @@
+	flash_cdns_pages_layout.pages_count =
+		(nand_param->npages_per_block * nand_param->nblocks_per_lun * nand_param->nluns);
+	flash_cdns_pages_layout.pages_size = nand_param->page_size;
+	*layout = &flash_cdns_pages_layout;
+	*layout_size = 1;
```
It does not seem you are querying this information from external device, so I guess you could just pre-calculate this at init, instead of doing that for each layout request.

de-nordic commented on drivers/flash/flash_cadence_nand.c:
```diff
@@ -0,0 +1,292 @@
+	ret = cdns_nand_write(nand_param, data, offset, len);
+
+	if (ret != 0) {
+		k_mutex_unlock(&nand_data->nand_mutex);
```
Same as in erase

nbalabak replied:
@de-nordic as per review comments changes updated

de-nordic commented on drivers/flash/flash_cadence_nand.c:
```diff
@@ -0,0 +1,292 @@
+	ret = cdns_nand_read(nand_param, data, offset, len);
+
+	if (ret != 0) {
+		k_mutex_unlock(&nand_data->nand_mutex);
```
same as in erase.

nbalabak replied:
@de-nordic as per review comments changes updated

de-nordic commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1524 @@
+ * @param params The Cadence NAND parameters structure.
+ * @retval 0 on success or negative error value on failure.
+ */
+#ifdef CONFIG_CDNS_NAND_INTERRUPT_SUPPORT
+static int cdns_transfer_complete(struct cdns_cdma_command_descriptor *desc_ptr,
+				  struct cadence_nand_params *params)
+#else
+static int cdns_transfer_complete(struct cdns_cdma_command_descriptor *desc_ptr)
+#endif
+{
```
Why just not define single cdsn_transfer_complete, with `struct cadence_nand_params *params` and do `ARG_UNUSED(params)` when not used?

nbalabak commented on drivers/flash/flash_cadence_nand.c:
```diff
@@ -0,0 +1,292 @@
+	ret = cdns_nand_erase(nand_param, offset, len);
+
+	if (ret != 0) {
+		k_mutex_unlock(&nand_data->nand_mutex);
```
@de-nordic as per review comments changes updated

nbalabak commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1519 @@
+ * @param thread The thread identifier for the PIO operation.
+ * @param vol_id The volume identifier for the feature set operation.
+ * @param use_intr Flag indicating whether to use interrupts during the operation.
+ * @retval  0 on success or negative error value on failure.
```
@de-nordic i updated changes

nbalabak commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1519 @@
+ *
+ * @param base_address The base address of the Cadence controller.
+ * @param thread The thread identifier to wait for.
+ * @retval  0 on success or negative error value on failure.
```
@de-nordic i updated changes

nbalabak commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1519 @@
+ */
+static int cdns_wait_for_thread(uintptr_t base_address, uint8_t thread)
+{
+#define THREAD_IDLE_TIME_OUT 500U
```
@de-nordic i updated changes

nbalabak commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1519 @@
+		return -EINVAL;
+	}
+	if ((size == 0) || (offset >= params->device_size) ||
+	    (((offset + size) - 1) >= params->device_size)) {
```
@de-nordic i updated changes

nbalabak commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1519 @@
+
+	if ((len == 0) || (offset >= params->device_size) ||
+	    (((offset + len) - 1) >= params->device_size)) {
+		LOG_ERR("Wrong offset value passed!!");
```
@de-nordic i updated changes

nbalabak commented on drivers/flash/flash_cadence_nand_ll.h:
```diff
@@ -0,0 +1,370 @@
+#endif
+} __aligned(32);
+
+/* Global function Api */
+int cdns_nand_init(struct cadence_nand_params *params);
+int cdns_nand_read(struct cadence_nand_params *params, const void *buffer, uint32_t offset,
+										uint32_t size);
+int cdns_nand_write(struct cadence_nand_params *params, const void *buffer, uint32_t offset,
+										uint32_t len);
+int cdns_nand_erase(struct cadence_nand_params *params, uint32_t offset, uint32_t size);
```
@de-nordic these api exposed to upper layer. we made two layers flash_cadence_nand_ll.c and  flash_cadence_nand.c.
 these function exposed to  flash_cadence_nand.c .

nbalabak commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1519 @@
+	if (opr_mode == CNF_OPR_WORK_MODE_SDR) {
+
+		status = ONFI_TIMING_MODE_SDR(
+			sys_read32(CNF_CTRLPARAM(base_address, ONFI_TIMING_0)));
+		timing_mode = find_lsb_set(status) - 1;
+
+		/* PHY Register Timing setting*/
+		sys_write32(PHY_CTRL_REG_SDR, (base_address + PHY_CTRL_REG_OFFSET));
+		sys_write32(PHY_TSEL_REG_SDR, (base_address + PHY_TSEL_REG_OFFSET));
+		sys_write32(PHY_DQ_TIMING_REG_SDR, (base_address + PHY_DQ_TIMING_REG_OFFSET));
+		sys_write32(PHY_DQS_TIMING_REG_SDR, (base_address + PHY_DQS_TIMING_REG_OFFSET));
+		sys_write32(PHY_GATE_LPBK_CTRL_REG_SDR, (base_address + PHY_GATE_LPBK_OFFSET));
+		sys_write32(PHY_DLL_MASTER_CTRL_REG_SDR, (base_address + PHY_DLL_MASTER_OFFSET));
+
+		/* Async mode timing settings */
+		sys_write32((CNF_ASYNC_TIMINGS_TRH) | (CNF_ASYNC_TIMINGS_TRP) |
+				    (CNF_ASYNC_TIMINGS_TWH) | (CNF_ASYNC_TIMINGS_TWP),
+			    CNF_MINICTRL(base_address, ASYNC_TOGGLE_TIMINGS));
+
+		/* Set operation work mode in common settings */
+		sys_clear_bits(CNF_MINICTRL(base_address, CMN_SETTINGS),
+			       CNF_OPR_WORK_MODE_SDR_MASK);
+
+	} else if (opr_mode == CNF_OPR_WORK_MODE_NVDDR) {
+
+		status = ONFI_TIMING_MODE_NVDDR(
+			sys_read32(CNF_CTRLPARAM(base_address, ONFI_TIMING_0)));
+		timing_mode = find_lsb_set(status) - 1;
+		/* PHY Register Timing setting*/
+		sys_write32(PHY_CTRL_REG_DDR, (base_address + PHY_CTRL_REG_OFFSET));
+		sys_write32(PHY_TSEL_REG_DDR, (base_address + PHY_TSEL_REG_OFFSET));
+		sys_write32(PHY_DQ_TIMING_REG_DDR, (base_address + PHY_DQ_TIMING_REG_OFFSET));
+		sys_write32(PHY_DQS_TIMING_REG_DDR, (base_address + PHY_DQS_TIMING_REG_OFFSET));
+		sys_write32(PHY_GATE_LPBK_CTRL_REG_DDR, (base_address + PHY_GATE_LPBK_OFFSET));
+		sys_write32(PHY_DLL_MASTER_CTRL_REG_DDR, (base_address + PHY_DLL_MASTER_OFFSET));
+		/* Set operation work mode in common settings */
+		sys_set_bits(CNF_MINICTRL(base_address, CMN_SETTINGS),
+			     CNF_OPR_WORK_MODE_NVDDR_MASK);
+
+	} else {
```
This code is not repeated; its behavior changes based on the SDR and DDR mode register values. The data rate is extracted from the dtsi, but there may be multiple nodes, some with SDR and others with DDR. Therefore, conditional compilation based on these scenarios is not feasible, in my opinion. Do you have any suggestions? Please share your insights

nbalabak replied:
> Looks like significant portion of code here repeats.

This code is not repeated; its behavior changes based on the SDR and DDR mode register values. The data rate is extracted from the dtsi, but there may be multiple nodes, some with SDR and others with DDR. Therefore, conditional compilation based on these scenarios is not feasible, in my opinion. Do you have any suggestions? Please share your insights

nbalabak replied:
> `opr_mode` check that could result in `-ENOTSUP`, and `opr_mode` check does not depend on run-time info from dev, should happen somewhere above the line 222 as there are already a lot of ops called on a device before deciding that the driver will not support the op anyway. I can also see that this is internal utility function and `opr_mode`, when this function gets called, comes from `cadence_nand_params->datarate_mode`, which is initialized at compile time from DTS. This means that the valid data rate mode could probably be checked by driver init, against allowed values, or verified at compile time. Is it possible to pass incorrect value here?

@de-nordic changes updated. datarate_mode value verified in dtsi via yml file not possible to get -ENOTSUP so removed

nbalabak commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1519 @@
+{
+	uint32_t status;
+
+	if (!WAIT_FOR(((cdns_nand_get_thrd_status(base_address, thread)) != 0), IDLE_TIME_OUT,
+		      k_msleep(1))) {
+		LOG_ERR("Timed out waiting for thread status response");
+		return -ETIMEDOUT;
+	}
+
+	status = cdns_nand_get_thrd_status(base_address, thread);
```
if only F_CSTAT_COMP bit then we can continue WAIT_FOR loop. but once complete flag set the possiblity there read/write operation fail bit also to set (line 195 and 196). did you have any sugesstion please let me know

nbalabak commented on dts/bindings/flash_controller/cdns,nand.yaml:
```diff
@@ -0,0 +1,19 @@
+properties:
+  data-rate-mode:
+    type: int
+    required: true
```
updated with enum values

nbalabak commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1517 @@
+		sys_clear_bits(CNF_MINICTRL(base_address, CMN_SETTINGS),
+			       CNF_OPR_WORK_MODE_SDR_MASK);
+
+	} else if (opr_mode == CNF_OPR_WORK_MODE_NVDDR) {
```
removed else if and changes to else

nbalabak commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1517 @@
+
+	base_address = params->nand_base;
+
+	/*Read flash device version information*/
```
changes updated

nbalabak commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1517 @@
+		return ret;
+	}
+
+	/*check device type*/
```
changes updated

nbalabak commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1517 @@
+		return ret;
+	}
+
+	/*DMA Setting*/
```
changes updated

nbalabak commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1517 @@
+	sys_write32((F_BURST_SEL_SET(NF_TDEF_BURST_SEL)) | (BIT(F_OTE)),
+		    (base_address + NF_DMA_SETTING));
+
+	/*Pre fetch*/
```
changes updated

nbalabak commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1517 @@
+	/* desc_ptr address passing */
+	sys_write32(desc_address & U32_MASK_VAL, (base_address + CDNS_CMD_REG2));
+	sys_write32((desc_address >> 32) & U32_MASK_VAL, (base_address + CDNS_CMD_REG3));
+	/*Thread selection*/
```
changes updated

nbalabak commented:
@de-nordic could you please review PR i fixed your review comments?

nbalabak commented:
> @de-nordic could you please review PR i fixed your review comments?

Hello @de-nordic,
       could you please review PR i fixed your review comments?

nbalabak commented:
> @povergoing could you please review PR i fixed your review comment?

@povergoing could you please review PR i fixed your review comment?

nbalabak commented:
@kartben could you please review this PR?

de-nordic commented on drivers/flash/flash_cadence_nand.c:
```diff
@@ -0,0 +1,287 @@
+	struct cadence_nand_params *nand_param = &nand_data->params;
+	int ret;
+
+	if (len == 0) {
+		LOG_ERR("Invalid input parameter for Nand Flash Erase!");
+		return -EINVAL;
+	}
```
Just do nothing, I think that is what most other drivers do.
I do not think that asking to erase 0 size is a failure, nothing really happens.

nbalabak replied:
removed this condition.

de-nordic commented on drivers/flash/flash_cadence_nand.c:
```diff
@@ -0,0 +1,287 @@
+	k_mutex_unlock(&nand_data->nand_mutex);
+
+	if (ret != 0) {
+		LOG_ERR("Cadence Nand Flash Erase Failed!");
```
I think it is NAND not Nand, and you could log the error here, in case it is -EINVAL that is coming somewhere from `cdns_nand_erase`, to distinguish from the one in line 92 . Bang is not needed, people are probably already scared by the "E:" at the beginning of error message anyway.

nbalabak replied:
removed this log error.

de-nordic commented on drivers/flash/flash_cadence_nand.c:
```diff
@@ -0,0 +1,287 @@
+	int ret;
+
+	if (len == 0) {
+		LOG_ERR("Invalid input parameter for Nand Flash Erase!");
```
I think this message is pointless. This is single point in `flash_cdns_nand_erase` where you can get the -EINVAL.
I think the `LOG_ERR` should be reserved for something that is coming from a lower level, that can not be easily identified.
I am quite sure that this can be easily figured out by user without the message.

de-nordic commented on drivers/flash/flash_cadence_nand.c:
```diff
@@ -0,0 +1,287 @@
+#endif
+
+#if CDNS_NAND_RESET_SUPPORT
+	/* Reset Combo phy and Nand only if reset controller driver is supported */
```
It is NAND.

nbalabak replied:
chnages updated.

de-nordic reviewed (commented):
Still reviewing...

de-nordic commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1516 @@
+					     CONFIG_FLASH_CDNS_CDMA_PAGE_COUNT, CNF_CMD_WR,
+					     start_page_number, buffer);
+
+	if (ret != 0) {
+		k_free(cdma_desc);
+		return ret;
+	}
```
Pointless, just return `ret` at the end.

nbalabak replied:
removed this condition

de-nordic commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1516 @@
+	struct cdns_cdma_command_descriptor *cdma_desc;
+	int ret;
+
+	cdma_desc = cdma_allocate_desc_memory(page_count, CONFIG_FLASH_CDNS_CDMA_PAGE_COUNT);
+
+	if (cdma_desc == NULL) {
+		LOG_ERR("Memory allocation error occurred %s", __func__);
+		return -ENOSR;
+	}
```
As far as I can tell, this could be done inside `cdns_cdma_desc_transfer_finish`.
Correct me if I am wrong but the first param to `cdma_allocate_desc_memory` is always third param to `cdns_cdma_desc_transfer_finish` and second param to the former is always fourth param to the later.
Wouldn't it save flash to move the block above to the `cdns_cdma_desc_transfer_finish` ?

nbalabak replied:
moved cdma_allocate_desc_memory functionality to cdns_cdma_desc_transfer_finish block.

de-nordic commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1516 @@
+					     CONFIG_FLASH_CDNS_CDMA_PAGE_COUNT, CNF_CMD_RD,
+					     start_page_number, buffer);
+
+	if (ret != 0) {
+		k_free(cdma_desc);
+		return ret;
+	}
+
+	k_free(cdma_desc);
```
same as in `cdns_nand_cdma_write`

nbalabak replied:
Changes updated

de-nordic commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1516 @@
+					     CONFIG_FLASH_CDNS_CDMA_BLOCK_COUNT, CNF_CMD_ERASE,
+					     start_block_number, NULL);
+
+	if (ret != 0) {
+		k_free(cdma_desc);
+		return ret;
+	}
+
+	k_free(cdma_desc);
```
Same as in `cdns_nand_cdma_write`

nbalabak replied:
Changes updated

de-nordic commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1516 @@
+		status |= CMD_0_C_MODE_SET(CT_PIO_MODE);
+		status |= PIO_CMD0_CT_SET(ctype);
+		sys_write32(status, (base_address + CDNS_CMD_REG0));
+#ifdef CONFIG_CDNS_NAND_INTERRUPT_SUPPORT
+		k_sem_take(&params->interrupt_sem_t, K_FOREVER);
+#endif
```
You could just have some `INT_SEM_TAKE` macro defined, that would depend on the `CONFIG_CDNS_NAND_INTERRUPT_SUPPORT`... less ifdefs cutting into code.

nbalabak replied:
macro added and replaced with new macro.

de-nordic commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1516 @@
+		return NULL;
+	}
+
+	memset(cdma_desc, 0, sizeof(struct cdns_cdma_command_descriptor) * page_count);
```
IMHO zeroing buffer at allocation is a waste of time, unless hardware assisted, because most of the buffer will be filled in with some stuff. Unless these buffers are passed to something that always requires it zeroed, then fine.

nbalabak replied:
removed.

de-nordic commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1516 @@
+					     CONFIG_FLASH_CDNS_CDMA_PAGE_COUNT, CNF_CMD_RD,
+					     start_page_number, buffer);
+
+	if (ret != 0) {
+		k_free(cdma_desc);
+		return ret;
+	}
```
There is double free everywhere where you do that after calling `cdns_cdma_desc_transfer_finish`, because on error conditions in `cdns_cdma_desc_transfer_finish` you release the buffer first time.

de-nordic commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1516 @@
+		row_address_set(params, &gen_row_address, start_page_number++);
+		address = ((uint64_t)gen_row_address);
+		if (mode == CDNS_READ) {
+			ret = cdns_generic_page_read(params, address,
```
I guess when `ret` is error you keep on looping?

nbalabak replied:
added return condition if error occurs.

de-nordic commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1516 @@
+ * @param data_length The length of the data to be copied.
+ * @param dir The direction of the copy operation (read or write).
+ */
+static void sdma_buffer_copy(uint32_t sdma_base_address, uint8_t *buffer, uint32_t data_length,
+			     uint8_t dir)
+{
+	uint32_t index;
+
+	if (dir == CDNS_READ) {
+		for (index = 0; index < data_length; index++) {
+			*(buffer + index) = sys_read8(sdma_base_address + index);
+		}
+	} else {
+		for (index = 0; index < data_length; index++) {
+			sys_write8(*(buffer + index), sdma_base_address + index);
+		}
+	}
+}
```
Not insisting but you could probably have just copyin/copyout fuctions for that.
I doubt we save any flash by stacking both directions here, assuming that the dir has to be passed to a function which means that each call has to pass it through some register, so probably save contents of the register on stack, and so on.

nbalabak replied:
copyin and copyout two seprate function added.

nbalabak replied:
@de-nordic  i fixed all review comments could you please review once?

de-nordic commented on drivers/flash/flash_cadence_nand.c:
```diff
@@ -0,0 +1,287 @@
+	struct cadence_nand_params *nand_param = &nand_data->params;
+	int ret;
+
+	if ((data == NULL) || (len == 0)) {
```
Now I see that the len is even checked in ll functions. Isn't there too much checking done, I mean on too many levels?

nbalabak replied:
removed from .c file len checking condition. Now ll function if len = 0 means it will return 0.

nbalabak commented on drivers/flash/flash_cadence_nand.c:
```diff
@@ -0,0 +1,287 @@
+		LOG_ERR("Cadence Nand Flash Erase Failed!");
+	}
+
+	return ret;
```
yes its in the ret range

nbalabak commented on drivers/flash/flash_cadence_nand_ll.c:
```diff
@@ -0,0 +1,1516 @@
+	}
+
+	params->nluns = CNF_GET_NLUNS(reg_value);
+	LOG_INF("Number of LUMs %x", params->nluns);
```
Changes updated

nbalabak commented:
@kartben @SgrrZhf @povergoing could you please review nand driver?

cfriedt reviewed (approved):
LGTM

nbalabak commented:
@de-nordic  could you merge my PR?

kartben commented:
> @de-nordic could you merge the PR?

One of our release engineers will probably pick it up shortly now that it has all required approvals.

nbalabak commented:
> > @de-nordic could you merge the PR?
> 
> One of our release engineers will probably pick it up shortly now that it has all required approvals.

@kartben Thank you for sharing information.

de-nordic commented:
> @de-nordic could you merge the PR?

I need to take my merge button for repair because it is gray and broken all the time. You will need to wait for release engineers to step in.
